packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1541,72 +1270,332 @@ module.exports = function<T, P, I, TI, PI, C, CC, CX, PL>(
+      // This root is not already scheduled. Add it.
+      root.remainingWork = expirationTime;
+      if (lastScheduledRoot === null) {
+        firstScheduledRoot = lastScheduledRoot = root;
+      } else {
+        lastScheduledRoot.nextScheduledRoot = root;
+        lastScheduledRoot = root;
+      }
+    } else {
+      // This root is already scheduled, but its priority may have increased.
+      const remainingWork = root.remainingWork;
+      if (remainingWork === null || expirationTime < remainingWork) {
+        // Update the priority.
+        root.remainingWork = expirationTime;
+      }
+    }
+
+    // If we're not already rendering, schedule work to flush now (if it's
+    // sync) or later (if it's async).
+    if (!isFlushingWork) {
+      // TODO: Remove distinction between sync and task. Maybe we can remove
+      // these magic numbers entirely by always comparing to the current time?
+      if (expirationTime === Sync) {
+        if (isUnbatchingUpdates) {
+          flushWork(Sync, null);
+        } else {
+          flushWork(Task, null);
+        }
+      } else if (!isCallbackScheduled) {
+        isCallbackScheduled = true;
+        scheduleDeferredCallback(flushAsyncWork);
+      }
+    }
+  }
+
+  function findHighestPriorityRoot() {
+    let highestPriorityWork = null;
+    let highestPriorityRoot = null;
+
+    let previousScheduledRoot = null;
+    let root = firstScheduledRoot;
+    while (root !== null) {
+      const remainingWork = root.remainingWork;
+      if (remainingWork === null) {
+        // If this root no longer has work, remove it from the scheduler.
+        let next = root.nextScheduledRoot;
+        root.nextScheduledRoot = null;
+        if (previousScheduledRoot === null) {
+          firstScheduledRoot = next;
+        } else {
+          previousScheduledRoot.nextScheduledRoot = next;
+        }
+        if (next === null) {
+          lastScheduledRoot = null;
+        }
+        root = next;
+        continue;
+      } else if (
+        highestPriorityWork === null ||
+        remainingWork < highestPriorityWork
+      ) {
+        // Update the priority, if it's higher
+        highestPriorityWork = remainingWork;
+        highestPriorityRoot = root;
+      }
+      previousScheduledRoot = root;
+      root = root.nextScheduledRoot;
+    }
+
+    // If the next root is the same as the previous root, this is a nested
+    // update. To prevent an infinite loop, increment the nested update count.
+    const previousFlushedRoot = nextFlushedRoot;
+    if (
+      previousFlushedRoot !== null &&
+      previousFlushedRoot === highestPriorityRoot
+    ) {
+      nestedUpdateCount++;
+    } else {
+      // Reset whenever we switch roots.
+      nestedUpdateCount = 0;
+    }
+    nextFlushedRoot = highestPriorityRoot;
+    nextFlushedExpirationTime = highestPriorityWork;
+  }
+
+  function flushAsyncWork(dl) {
+    flushWork(null, dl);
+  }
+
+  function flushWork(
```

Comment:
Let's call this something other than `flush` since it might not commit side-effects. E.g. `perform`.